state-sync/storage-service/types/src/lib.rs

```diff
@@ -495,14 +495,12 @@ impl<T: PrimInt> CompleteDataRange<T> {
     }
 
     /// Returns the length of the data range.
-    ///
-    /// Note that `CompleteDataRange` maintains the invariant that the range
-    /// length is always expressible without overflow, so this method is
-    /// infallible.
     #[inline]
-    pub fn len(&self) -> T {
-        debug_assert!(range_length_checked(self.lowest, self.highest).is_ok());
-        (self.highest - self.lowest) + T::one()
+    pub fn len(&self) -> Result<T, DegenerateRangeError> {
+        match range_length_checked(self.lowest, self.highest) {
+            Ok(_) => Ok((self.highest - self.lowest) + T::one()),
```

Comment:
Instead of using the binary operators `-` and `+` that have the potential for overflows, the aim here is to use checked arithmetic that returns an error when an overflow occurs, e.g., see here: https://github.com/diem/diem/blob/8b71992310aff321d0876c9b4da18c65ed1e64b2/state-sync/state-sync-v1/src/coordinator.rs#L1005.

Perhaps you could create a new PR following this that goes through all `diem/state-sync/storage-service/*` code and replaces these types of operators with checked arithmetic? We should already be doing it for some operators. Likewise with the `diem/state-sync/diem-data-client/*` files 😄

Reply:
Hi @JoshLind , `range_length_checked` does seem to already check the overflow using the checked arithmetic. I just want to clarify if you mean: (1) we still use/keep `match range_length_checked`, but we'd still like to replace `+/-` inside `Ok` arm with checked arithmetic (though it's impossible for overflow to happen in this case)?, or (2) simply remove `match range_length_checked` and calculate the length directly using checked arithmetic instead?

Reply:
Used checked arithmetic operations to calculate range length.

Reply:
Aah, I see what you mean. Yeah, sometimes it's annoying to use checked arithmetic when we know an overflow isn't possible, but my generally approach is to do it anyway because otherwise when we run static analysis tools to look for these problems, the tools aren't smart enough to know, so they produce false-positives. In my mind, if we always do it, we're not missing any cases 😄

Reply:
Noted. Thanks for the tips! I'd pay attention to this when writing Rust code in future.